src/main/java/com/amazonaws/services/dynamodbv2/datamodeling/encryption/DynamoDBEncryptor.java

```diff
@@ -539,17 +558,22 @@ public String getSigningAlgorithmHeader() {
             attributeValue.getB().reset();
         }
     }
-    
+
     private static byte[] toByteArray(ByteBuffer buffer) {
-        if (buffer.hasArray()) {
+        buffer = buffer.duplicate();
+        // We can only return the array directly if:
+        // 1. The ByteBuffer exposes an array
+        // 2. The ByteBuffer starts at the beginning of the array
+        // 3. The ByteBuffer uses the entire array
+        if (buffer.hasArray() && buffer.arrayOffset() == 0) {
             byte[] result = buffer.array();
-            buffer.rewind();
-            return result;
-        } else {
-            byte[] result = new byte[buffer.remaining()];
-            buffer.get(result);
-            buffer.rewind();
-            return result;
+            if (buffer.remaining() == result.length) {
```

Comment:
Combine into the if condition.

Reply:
Can't do as `result` can only be safely retrieved from `buffer.array()` after the first conditional and we want to check it for sanity prior to returning it.

Reply:
got it, the `byte[] result` assignment was lost between two diff blocks.